bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.openthermgateway.internal;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.AbstractMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.openthermgateway.handler.TypeConverter;
+
+/**
+ * The {@link OpenThermGatewaySocketConnector} is responsible for handling the socket connection
+ *
+ * @author Arjen Korevaar - Initial contribution
+ * @author Arjan Mels - Improved robustness by re-sending commands, handling all message types (not only Boiler)
+ */
+@NonNullByDefault
+public class OpenThermGatewaySocketConnector implements OpenThermGatewayConnector {
+    private static final int COMMAND_RESPONSE_TIME = 100;
+    private static final int COMMAND_TIMEOUT = 5000;
+
+    private OpenThermGatewayCallback callback;
+    private String ipaddress;
+    private int port;
+
+    private @Nullable Socket socket;
+    private @Nullable BufferedReader reader;
+    private @Nullable PrintWriter writer;
+
+    private volatile boolean stopping;
+    private boolean connected;
+
+    public OpenThermGatewaySocketConnector(OpenThermGatewayCallback callback, String ipaddress, int port) {
+        this.callback = callback;
+        this.ipaddress = ipaddress;
+        this.port = port;
+    }
+
+    @Override
+    public void run() {
+        stopping = false;
+        connected = false;
+
+        try {
+            callback.log(LogLevel.Info,
+                    String.format("Connecting OpenThermGatewaySocketConnector to %s:%s", this.ipaddress, this.port));
+
+            callback.connecting();
+
+            socket = new Socket();
+            socket.connect(new InetSocketAddress(this.ipaddress, this.port), COMMAND_TIMEOUT);
+            socket.setSoTimeout(COMMAND_TIMEOUT);
+            writer = new PrintWriter(socket.getOutputStream(), true);
+            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
```

Comment:
Use try-with-resources:
```
try (writer = new PrintWriter(socket.getOutputStream(), true);
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
```
Then you don't have to call close.

Reply:
Like the using statement from c#, didn't know Java had such a thing. Anyway, causes compiler errors on my machine, so need to look into this..

Reply:
Fixed compiler error with updating java, implemeted try-with-resources